### PR TITLE
Add elastic beat shippers

### PR DESCRIPTION
**Pull request by fadenb**
###### Motivation for this change
Packages 4 of the Elastic Beats shippers
https://www.elastic.co/products/beats

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/misc/logging/filebeat/default.nix [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    mkdir -p $out/bin
+    cp filebeat $out/bin/
+  '';
+
+  meta = {
+    description = "Lightweight shipper for logfiles";
+    homepage = https://www.elastic.co/products/beats;
+    license = stdenv.lib.licenses.asl20;
+    maintainers = [ stdenv.lib.maintainers.fadenb ];
+    platforms = stdenv.lib.platforms.all;
```
Comment (fpletz) [inline pkgs/misc/logging/filebeat/default.nix]: Only `"linux-x86_64"` is supported.

File: pkgs/misc/logging/heartbeat/default.nix [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    mkdir -p $out/bin
+    cp heartbeat $out/bin/
+  '';
+
+  meta = {
+    description = "Lightweight shipper for uptime monitoring";
+    homepage = https://www.elastic.co/products/beats;
+    license = stdenv.lib.licenses.asl20;
+    maintainers = [ stdenv.lib.maintainers.fadenb ];
+    platforms = stdenv.lib.platforms.all;
```
Comment (fpletz) [inline pkgs/misc/logging/heartbeat/default.nix]: Only `"linux-x86_64"` is supported.

File: pkgs/misc/logging/metricbeat/default.nix [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    mkdir -p $out/bin
+    cp metricbeat $out/bin/
+  '';
+
+  meta = {
+    description = "Lightweight shipper for metrics";
+    homepage = https://www.elastic.co/products/beats;
+    license = stdenv.lib.licenses.asl20;
+    maintainers = [ stdenv.lib.maintainers.fadenb ];
+    platforms = stdenv.lib.platforms.all;
```
Comment (fpletz) [inline pkgs/misc/logging/metricbeat/default.nix]: Only `"linux-x86_64"` is supported.

File: pkgs/misc/logging/packetbeat/default.nix [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    longDescription = ''
+      Packetbeat is an open source network packet analyzer that ships the data to Elasticsearch.
+
+      Think of it like a distributed real-time Wireshark with a lot more analytics features.
+      The Packetbeat shippers sniff the traffic between your application processes, parse on the fly protocols like HTTP, MySQL, PostgreSQL, Redis or Thrift and correlate the messages into transactions.
+    '';
+    homepage = https://www.elastic.co/products/beats;
+    license = stdenv.lib.licenses.asl20;
+    maintainers = [ stdenv.lib.maintainers.fadenb ];
+    platforms = stdenv.lib.platforms.all;
```
Comment (fpletz) [inline pkgs/misc/logging/packetbeat/default.nix]: Only `"linux-x86_64"` is supported.

**Review by fpletz — CHANGES_REQUESTED** [review]
I'm wondering if we can just compile these packages? They're open source after all. :smile:

File: pkgs/misc/logging/metricbeat/default.nix [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  };
+
+  # statically linked binary, no need to build anything
+  dontBuild = true;
+  doCheck = false;
+
+  # need to patch interpreter to be able to run on NixOS
+  patchPhase = ''
+    patchelf --interpreter $(cat $NIX_CC/nix-support/dynamic-linker) metricbeat
+  '';
```
Comment (fpletz) [inline pkgs/misc/logging/metricbeat/default.nix]: Why does this package need this but not the others? Shouldn't the default `fixupPhase` take care of that?

**Comment by fadenb** [comment]
Last time I tried building the beats from source I failed so this was the next best thing to get them working on NixOS ;)

Regarding the patchelf command:
Two of the four beats require patchelf to be run on them in order to work on NixOS. Seems like it is not done by default.

**Review by fpletz — APPROVED** [review]

**Comment by danbst** [comment]
Last time I tried Filebeat, this compiled and was runnable:
```
{ fetchurl, buildGoPackage }:

buildGoPackage rec {
  name = "filebeat-${version}";
  goPackagePath = "github.com/elastic/beats";
  version = "5.1.2";
  src = fetchurl {
    url = "https://${goPackagePath}/archive/v${version}.tar.gz";
    sha256 = "1ki37sm6yj7m1ixfsb37128wagfymhh9hz61pc72j0kbpvw59mbw";
  };

  subPackages = [ "filebeat" ];
}
```
Unfortunately it propagates src into derivation, so requires polishing.

**Comment by teh** [comment]
Though I can't offer much help I'm a light -1 on introducing more binary installs, even if that's a bit more effort. Source installs allow us to track the real build dependencies, and to apply patches or overrides later. It's also useful for reproducible builds.

For some packages we don't have a functioning build tool yet, e.g. bazel & tensorflow, or some maven packages. In this specific instance though I think all that's needed is a bit of fiddling.

E.g. I run a go project with a custom Makefile and my installer looks like this:

```
  buildPhase = ''
    shopt -s extglob
    mkdir -p ./src/github.com/coreos/dex
    mv !(src) ./src/github.com/coreos/dex
    GOPATH=$PWD
    cd ./src/github.com/coreos/dex
    make
  '';
  installPhase = ''
    mkdir -p $out/bin
    cp bin/dex $out/bin
  '';
```

**Review by fpletz — APPROVED** [review]

**Pull request merged**